cms/djangoapps/modulestore_migrator/api/read_api.py

```diff
@@ -0,0 +1,263 @@
+"""
+API for reading information about previous migrations
+"""
+from __future__ import annotations
+
+import typing as t
+from uuid import UUID
+
+from opaque_keys.edx.keys import UsageKey
+from opaque_keys.edx.locator import (
+    CourseLocator, LibraryLocator,
+    LibraryLocatorV2, LibraryUsageLocatorV2, LibraryContainerLocator
+)
+from openedx_learning.api.authoring import get_draft_version
+from openedx_learning.api.authoring_models import (
+    PublishableEntityVersion, PublishableEntity, DraftChangeLogRecord
+)
+
+from openedx.core.djangoapps.content_libraries.api import (
+    library_component_usage_key, library_container_locator
+)
+
+from ..data import (
+    SourceContextKey, ModulestoreMigration, ModulestoreBlockMigrationResult,
+    ModulestoreBlockMigrationSuccess, ModulestoreBlockMigrationFailure
+)
+from .. import models
+
+
+__all__ = (
+    'forward_course',
+    'forward_legacy_library',
+    'forward_context',
+    'forward_blocks',
+    'get_migrations',
+    'get_migration_blocks',
+)
+
+
+def forward_blocks(source_keys: t.Iterable[UsageKey]) -> dict[UsageKey, ModulestoreBlockMigrationSuccess]:
+    """
+    Authoritatively determine how some Modulestore blocks have been migrated to Learning Core.
+
+    Returns a mapping from source usage keys to block migration data objects. Each block migration object
+    holds the target usage key and title. If a source key is missing from the mapping, then it has not
+    been authoritatively migrated.
+    """
+    sources = models.ModulestoreBlockSource.objects.filter(
+        key__in=[str(sk) for sk in source_keys]
+    ).select_related(
+        "forwarded__target__learning_package",
+        # For building component key
+        "forwarded__target__component__component_type",
+        # For building container key
+        "forwarded__target__container__section",
+        "forwarded__target__container__subsection",
+        "forwarded__target__container__unit",
```

Comment:
I don't love hard-coding these 3 specific container types into the select_related here, but I think it's the only way to build the key without executing an additional query.

Reply:
Put that in an inline comment?